internal/schema/connection.go

```diff
@@ -3,9 +3,128 @@ package schema
 import (
 	"fmt"
 
+	"github.com/hashicorp/terraform-plugin-framework-validators/listvalidator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	fwschema "github.com/hashicorp/terraform-plugin-framework/provider/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
+func GetFWConnectionBlock(keyName string, isProviderConfiguration bool) fwschema.Block {
+	usernamePath := makePathRef(keyName, "username")
+	passwordPath := makePathRef(keyName, "password")
+	caFilePath := makePathRef(keyName, "ca_file")
+	caDataPath := makePathRef(keyName, "ca_data")
+	certFilePath := makePathRef(keyName, "cert_file")
+	certDataPath := makePathRef(keyName, "cert_data")
+	keyFilePath := makePathRef(keyName, "key_file")
+	keyDataPath := makePathRef(keyName, "key_data")
+
+	usernameValidators := []validator.String{stringvalidator.AlsoRequires(path.MatchRoot(passwordPath))}
+	passwordValidators := []validator.String{stringvalidator.AlsoRequires(path.MatchRoot(usernamePath))}
+
+	if isProviderConfiguration {
+		// RequireWith validation isn't compatible when used in conjunction with DefaultFunc
+		usernameValidators = nil
+		passwordValidators = nil
+	}
+
+	return fwschema.ListNestedBlock{
+		MarkdownDescription: fmt.Sprintf("Elasticsearch connection configuration block. %s", getDeprecationMessage(isProviderConfiguration)),
+		DeprecationMessage:  getDeprecationMessage(isProviderConfiguration),
+		NestedObject: fwschema.NestedBlockObject{
+			Attributes: map[string]fwschema.Attribute{
+				"username": fwschema.StringAttribute{
+					MarkdownDescription: "Username to use for API authentication to Elasticsearch.",
+					Optional:            true,
+					Validators:          usernameValidators,
+				},
+				"password": fwschema.StringAttribute{
+					MarkdownDescription: "Password to use for API authentication to Elasticsearch.",
+					Optional:            true,
+					Sensitive:           true,
+					Validators:          passwordValidators,
+				},
+				"api_key": fwschema.StringAttribute{
+					MarkdownDescription: "API Key to use for authentication to Elasticsearch",
+					Optional:            true,
+					Sensitive:           true,
+					Validators: []validator.String{
+						stringvalidator.ConflictsWith(path.MatchRoot(usernamePath)),
```

Comment:
Can we instead use a relative path matcher in the new schema? 

```suggestion
						stringvalidator.ConflictsWith(path.MatchRelative().AtParent().AtName("username")),
```